### PR TITLE
Refactor BurnMintERC677 to inherit from generic BurnMintERC20

### PR DESCRIPTION
**Pull request by jhweintraub**
The existing implementation of BurnMintERC677 inherits the ERC20 functionality from OZ, but not any of the BurnMint Functionality that is desired from it. As a result there exists no standard BurnMintERC20 implementation for Chainlink projects/customers to use without requiring they also support ERC-677 functionality

This PR restructures the `BurnMintERC677` implementation to inherit from a more generic `BurnMintERC20` and then adds the `transferAndCall` method and updates to `supportsInterface()` in to ensure compliance with ERC677. The existing base implementation of the standard `ERC677.sol` has not been modified.

### CODE REVIEW

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol [hunk 1]
```diff
@@ -1,217 +1,33 @@
...
-  function revokeMintRole(address minter) public onlyOwner {
-    if (s_minters.remove(minter)) {
-      emit MintAccessRevoked(minter);
+contract BurnMintERC677 is BurnMintERC20, IERC677 {
+  constructor(
+    string memory name,
+    string memory symbol,
+    uint8 decimals_,
+    uint256 maxSupply_
+  ) BurnMintERC20(name, symbol, decimals_, maxSupply_) {}
```
Comment (RyanRHall) [inline contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol]: nit: I think we can drop the `_`s here. In general, I think apply the campsite rule here "leave it better than you found it". This is old code that we're touching for the first time in a while, so it's nice to use the opportunity to bring the code quality up to par with our style guide :)

**Review by RyanRHall — CHANGES_REQUESTED** [review]

**Review by RyanRHall — DISMISSED** [review]
👏👏👏 this refactor is 💯  ... just one nit

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol (hunk 1 above)
Comment (jhweintraub) [reply contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol]: I thought the campsite rule was "leave no trace" in which case changing nothing would be key but sure I can change it no problem

**Review by 0xsuryansh — DISMISSED** [review]
LGTM!

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol [hunk 2]
```diff
@@ -1,217 +1,33 @@
...
-      emit BurnAccessGranted(burner);
-    }
-  }
-
-  /// @notice Revokes mint role for the given address.
-  /// @dev only the owner can call this function.
-  function revokeMintRole(address minter) public onlyOwner {
-    if (s_minters.remove(minter)) {
-      emit MintAccessRevoked(minter);
+contract BurnMintERC677 is BurnMintERC20, IERC677 {
```
Comment (RyanRHall) [inline contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol:11]: Can we inherit `ERC677` here so that we can avoid re-defining `transferAndCall()` below?

Comment (jhweintraub) [reply contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol:11]: No, because that means double importing ERC20 constructor which the compiler will not allow. The `ERC677.sol` implementation is base ERC20 + `transferAndCall`, the BurnMintERC20 is ERC20 + all the permission stuff we added. You cannot import both at the same time cause of the double constructor, and we want `ERC677.sol` to be a minimal implementation of ERC20 without all the extra stuff we added

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol [hunk 3]
```diff
@@ -1,217 +1,33 @@
...
+  constructor(
+    string memory name,
+    string memory symbol,
+    uint8 decimals,
+    uint256 maxSupply
+  ) BurnMintERC20(name, symbol, decimals, maxSupply) {}
+
+  /// @inheritdoc IERC677
+  function transferAndCall(address to, uint256 amount, bytes memory data) public returns (bool success) {
+    super.transfer(to, amount);
```
Comment (RyanRHall) [inline contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol]: nit: I don't think we need `super` here, right? I'm not sure why this is called this way (same in ERC677)

**Review by RyanRHall — DISMISSED** [review]

**Review by 0xsuryansh — DISMISSED** [review]

**Review by RyanRHall — DISMISSED** [review]

**Review by matYang — DISMISSED** [review]

File: contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol [hunk 4]
```diff
@@ -2,17 +2,16 @@
 pragma solidity 0.8.24;
 
 import {BurnMintERC677} from "../../../shared/token/ERC677/BurnMintERC677.sol";
-import {IGetCCIPAdmin} from "../../interfaces/IGetCCIPAdmin.sol";
 
-contract BurnMintERC677Helper is BurnMintERC677, IGetCCIPAdmin {
+contract BurnMintERC677Helper is BurnMintERC677 {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol]: Why was this removed?

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol [hunk 5]
```diff
@@ -1,217 +1,33 @@
-// SPDX-License-Identifier: MIT
```
Comment (RensR) [inline contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol:1]: Why are we changing this?

File: core/gethwrappers/shared/generated/link_token/link_token.go [hunk 6]
```diff
@@ -1932,11 +1932,11 @@ func (LinkTokenOwnershipTransferred) Topic() common.Hash {
 }
 
 func (LinkTokenTransfer) Topic() common.Hash {
-	return common.HexToHash("0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef")
+	return common.HexToHash("0xe19260aff97b920c7df27010903aeb9c8d2be5d310a2c67824cf3f15396e4c16")
```
Comment (RensR) [inline core/gethwrappers/shared/generated/link_token/link_token.go:1935]: What causes this diff? Can we undo that?

Comment (jhweintraub) [reply core/gethwrappers/shared/generated/link_token/link_token.go:1935]: auto generated by the CI gethwrappers

File: contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol (hunk 5 above)
Comment (jhweintraub) [reply contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol:1]: I thought you said you wanted it at BUSL license?

File: contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol (hunk 4 above)
Comment (jhweintraub) [reply contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol]: I probably removed it somewhere in the development process when it was causing inheritance issues and forgot to put it back in at some point. This is a helper contract does it matter?

File: core/gethwrappers/shared/generated/link_token/link_token.go (hunk 6 above)
Comment (RensR) [reply core/gethwrappers/shared/generated/link_token/link_token.go:1935]: I know that, but I'd like it to stay the same. Why would the hash change was my question, did something in this onchain logic change?

File: contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol (hunk 4 above)
Comment (RensR) [reply contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol]: Why would you remove it if it doesn't matter? This is deployed to testnets so yes it matters.

Comment (jhweintraub) [reply contracts/src/v0.8/ccip/test/helpers/BurnMintERC677Helper.sol]: I mean I had the opposite belief, that if it didn't matter then remove it for readability but thats fine it's been re-added anyways, this comment is stale anyways

File: core/gethwrappers/shared/generated/link_token/link_token.go (hunk 6 above)
Comment (jhweintraub) [reply core/gethwrappers/shared/generated/link_token/link_token.go:1935]: It shouldn't have. I will investigate.
